### DISCUSSION TITLE
More generic multimodality support

### DISCUSSION THREAD

**Discussion by Wojtab** [oobabooga/text-generation-webui#1687] [Ideas]
Since writing the LLaVA extension, I was thinking whether/how to add support for more multimodal models. I think there is a way to make a generic multimodal wrapper for [LLaVA](https://github.com/haotian-liu/LLaVA)/[MiniGPT-4](https://github.com/Vision-CAIR/MiniGPT-4), presumably also for [mPLUG-Owl](https://github.com/X-PLUG/mPLUG-Owl), and all of the models which only input `input_embeds` (so not ones like [OpenFlamingo](https://github.com/mlfoundations/open_flamingo), which adds cross-attention layers to LLM).

Now, the main assumption is that the models use the following "syntax" for their embeds:
```
<user text><(optional) image start token/text><N image embeddings><(optional) image stop token/text><user text>
```
which is the case for both LLaVA and MiniGPT-4:
||LLaVA|MiniGPT-4|
|-|-|-|
|image start token/text|`<im_start>` (token)|\<Img> (string)|
|image stop token/text|`<im_stop>` (token)|\</Img> (string)|
|N|256 * `<im_pad>` (token)|32 * None (just embeds, no placeholder)|

So the LLM part is very similar. Now, where they differ is the vision pipeline:
LLaVA: `CLIP -> Linear -> LLM`
MiniGPT-4: `ViT -> QFormer -> Linear -> LLM`
mPLUG-Owl: `modified CLIP -> mPLUG_OwlVisualAbstractorEncoder -> Linear -> LLM`

To support them, we'll need to add a small python script per model(ok, maybe for the owl it's not going to be small), which will download the models, and set up the pipeline, something similar to `LLaVAEmbedder`, but more bare bones, for example:
```Python
class AbstractMultimodalPipeline:
  def init():
    # download + load the models
  def image_start() -> str:
    # return image start string
  def image_end() -> str:
    # return image end string
  def num_image_embeds() -> int:
    # return the number of embeds used by a single image (256 for LLaVA)
  def embed_image(image) -> torch.Tensor:
    # forward the image through vision pipeline, and return the embeddings
```
Each concrete model would then inherit from this class, and be used by a generalized workflow. Now, there are 2 ways to go about it:
1. Each model uses a separate extension, I call it "TTS solution", following 5 different TTS extensions
    - Could be a module, from which the extensions will import, which will set up appropriate hooks/ui/deal with images etc.
2. There is one single extension, with all of the models, it then loads the appropriate one based on a CLI parameter, I think it's a better idea

Another question is if we should include so many configs in webui, just LLaVA and MiniGPT-4 need 3 (LLaVA 13B, MiniGPT-4 13b, MiniGPT-4 7b), with LLaVA planning to release 7b, and maybe Vicuna-1.1 based models makes it either 4 or 6, and probably many more coming(like this owl model). This is mainly a question for @oobabooga 

Feedback welcome

EDIT: I guess my timing was right, as LLaVA-7B got released an hour after I posted this

EDIT2: for a POC see https://github.com/oobabooga/text-generation-webui/pull/1741

**Comment by TomLucidor** [comment]
Secondary thought: are there open source multi-modal models? https://github.com/open-mmlab/Multimodal-GPT
